### DISCUSSION TITLE
Dynamic Page – React state carries over?

### DISCUSSION THREAD

**Discussion by taylorbourne** [vercel/next.js#15380] [Help] [answered]
Hey All,

I have a pretty basic site that uses dynamic routes pulled from a local data file. Basically I have...

```
/services (services/index.js)
/services/[service] (services/[service].js)
```

I'm using both `getStaticProps` and `getStaticPaths` in `services/[service].js` as recommended in the docs:

```
export const getStaticPaths = () => {
  const paths = Object.keys(serviceData.services).map((service) => ({
    params: { service },
  }))
  return { paths, fallback: false }
}

export const getStaticProps = () => {
  return {
    props: {
      pageData: serviceData,
    },
  }
}
```

Note – `serviceData` is an object imported from my `data.js` file which is in the root folder of the project. There's no API call being made here.

In `[service].js` I'm using Reacts `useState` to track a currently selected `activeSubService`, which is only populated (via `setActiveSubService('subService')`) when the user clicks on one of the "sub services" on the service page.

When I navigate from one dynamic `[service].js` route to another I noticed that if I had selected a sub service (thereby firing `setActiveSubService()`) the next route still has the active sub service set from the previous page. Is this expected? Currently I have to add some extra checks to make sure that the data isn't junk since the state is completely invalid.

Am I doing something wrong here?

Thanks!

**Comment by jamesmosier** [comment, marked as answer]
Hi @taylorbourne. It looks like this is currently the way it functions. I was able to reproduce this functionality and state was maintained between dynamic routes. 

There is an open issue with some suggestions: https://github.com/vercel/next.js/issues/9992

I'd say using something like `useEffect` to reset state upon changing routes would be a good option though.

**Reply by taylorbourne** [reply]
@jamesmosier thanks so much for the reply, really appreciate it. 

Bam, exactly what I needed to hear. I really just wanted to know that I’m not crazy, and I’m not just running into some quickly solvable user-error.  The workarounds provided in the link you posted should be more than enough to avoid the ugly conditionals I have and keep my code readable. 

This might be something that’s worth updating in the docs unless it’s unintentional. At the very least it should be added as a “caveat” of dynamic pages. At first, because the state is preset when you navigate to the second dynamic page, I thought that this was a bug with `getStaticProps` which would mean that the NextJS core had an underlying bug. The actual problem is much less to worry about, and could even be described as a feature.
